rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,377 @@
+| 400  | if the files type is invalid                                                                                            |
+
+### Upload Flows
+
+There are two methods for uploading data to the DP. They are Cloud to Cloud and Local to Cloud. The follow
+sections details the user interactions required, and how the Data Portal backend will facilitate the upload.
+
+#### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their CSP from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+##### 1. Get Share Link
+
+The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or submission of files.
+
+##### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery.
+
+###### How to download from a shared link
+
+- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
+- Google Drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>
+
+##### 4. _Optional_ Queue
+
+This part is optional and depends on the load expected and server type selected. The queue stores the links that need to
+be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
+It also gives us the ability to retry a download if one of the servers fails to download or upload the file.
+
+###### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- The CSP experience and outage
+- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+##### 5. Start EC2 Instance to Download File
+
+Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.
```

Comment:
I think we're going to want EBS-backed EC2 instances? I'd check with @mckinsel and/or the AWS team.